### PR TITLE
[DNM] use separated thread to service read only query

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
ref https://github.com/pingcap/tikv/issues/1617
I will introduce bench result for this branch later.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
deployment: 1PD 3TiKV 1TiDB, 16 tables, each table 1000,000 rows.
sysbench result:
```
branch-select
qps: 30910.42 per sec.
response avg: 2.07ms

master-select
qps: 30570.16 per sec.
response avg: 2.09ms

branch-insert
qps: 6673.61 per sec.
avg: 4.79ms

master-insert
qps: 6567.30 per sec.
avg: 4.87ms
```

File: src/raftstore/store/local_read.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
+
```
Comment (BusyJay) [inline src/raftstore/store/local_read.rs]: Missing licence.

File: src/raftstore/store/local_read.rs [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+            return false;
+        }
+
+        true
+    }
+}
+
+type Key = Vec<u8>;
+
+pub struct LocalReadHandler {
```
Comment (BusyJay) [inline src/raftstore/store/local_read.rs]: Why not use `Worker`?

File: src/raftstore/store/msg.rs [hunk 3]
```diff
@@ -74,6 +75,28 @@ pub enum Msg {
...
+        region: metapb::Region,
+        peer: metapb::Peer,
+        leader_id: u64,
+        term: u64,
+        applied_index_term: u64,
+    },
+
+    RemovePeerStatus { region_id: u64 },
+
+    UpdatePeerStatus(PeerStatusChange),
```
Comment (siddontang) [inline src/raftstore/store/msg.rs]: use struct field type, not tuple here

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -755,6 +777,18 @@ impl Peer {
                        current_expired_time,
                        next_expired_time);
                 self.leader_lease_expired_time = Some(Either::Left(next_expired_time));
+
+                // Notify local read thread to update lease lazily.
+                let notify_gap = Duration::seconds(1);
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: can we define a const var for it, or use config?

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -145,13 +150,16 @@ pub fn delete_file_in_range(db: &DB, start_key: &[u8], end_key: &[u8]) -> Result
 }
 
 impl<T, C> Store<T, C> {
+    #[allow(too_many_arguments)]
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: if too_many_arguments, we should use a Options instead.

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -578,14 +581,30 @@ impl Peer {
...
+                    let leader_lease_expired_time =
                         Some(Either::Left(self.next_lease_expired_time(clocktime::raw_now())));
+                    self.leader_lease_expired_time = leader_lease_expired_time.clone();
                     debug!("{} becomes leader and lease expired time is {:?}",
                            self.tag,
                            self.leader_lease_expired_time);
+                    local_read_tx.send(Msg::UpdatePeerStatus(PeerStatusChange {
+                            region_id: self.region().get_id(),
+                            leader_id: Some(ss.leader_id),
+                            leader_lease_expired_time: Some(leader_lease_expired_time.clone()),
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: no need clone here.

File: src/raftstore/store/local_read.rs (hunk 2 above)
Comment (siddontang) [inline src/raftstore/store/local_read.rs]: +1

File: src/raftstore/store/local_read.rs [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+    peers_status: HashMap<u64, PeerStatus>,
+    store_id: u64,
+    engine: Arc<DB>,
+
+    // used when need redirect request to raftstore
+    sendch: SendCh<Msg>,
+    redirect_cmds: HashMap<Uuid, Callback>,
+
+    // region end key -> region id
+    region_ranges: BTreeMap<Key, u64>,
```
Comment (siddontang) [inline src/raftstore/store/local_read.rs]: I think we don't need region_ranges here, if we meet the stale or lease expired error, we can all redirect the request to the raftstore thread.

File: src/raftstore/store/local_read.rs [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+                if self.region_ranges.insert(end_key, region_id).is_none() {
+                    panic!("[Override] range for region [{}] is not exist.", region_id);
+                }
+            }
+        }
+    }
+
+    fn redirect_cmd_to_raftstore(&mut self, uuid: Uuid, req: RaftCmdRequest, cb: Callback) {
+        self.redirect_cmds.insert(uuid, cb);
+        if let Err(e) = self.sendch.send(Msg::RedirectRaftCmd {
```
Comment (siddontang) [inline src/raftstore/store/local_read.rs]: I prefer send return the RedirectRaftCmd directly, so we don't need redirect_cmds.

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -1980,6 +2141,24 @@ impl<T: Transport, C: PdClient> mio::Handler for Store<T, C> {
...
+                self.propose_raft_command(request,
+                                          box move |resp| {
+                    local_read_tx.send(Msg::RedirectRaftCmdResp {
+                            uuid: uuid,
+                            resp: resp,
+                        })
+                        .expect("send RedirectRaftCmdResp to local read thread")
+                });
+            }
+            Msg::RedirectRaftCmdResp { .. } |
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: why not use `_ => unreachable!()` directly?

**Pull request closed without merging**